### PR TITLE
New analysis layer base code for scheduled processing

### PR DESCRIPTION
**Pull request by Sam-Hutchins**
RED team is interested in creating a scheduled process for showing primary statistics and demographics about students.  The key features are the student's school, teacher, courses started, and last level completed within a given school year.  We would like this process to run nightly in order to use in daily analysis as a source of truth.

## Testing story

We don't have a good way of testing that these changes are error-free but we have been doing UAT (on the RED team) for two weeks using a "work in progress" file and have not revealed any major defects.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: aws/redshift/tables/historic_students.sql [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+st.level_id as start_level_id,
+en.script_id as end_script_id,
+en.stage_id as end_stage_id,
+en.level_id as end_level_id
+from ul_start st
+join ul_end en on st.user_id = en.user_id and st.school_year = en.school_year and st.course_name = en.course_name
+join dashboard_production_pii.users u on st.user_id = u.id  and u.user_type = 'student'
+join dashboard_production.user_geos ug on u.id = ug.user_id and ug.country = 'United States'
+left join student_sections sts on st.user_id = sts.student_user_id and st.school_year = sts.school_year
+left join public.rosetta_historic_teachers ht on ht.teacher_user_id = sts.teacher_user_id and ht.school_year = st.school_year);
```
Comment (wjordan) [inline aws/redshift/tables/historic_students.sql:64]: this table seems to be missing- is there another script to generate this table that also runs on a scheduled basis?

**Review by wjordan — COMMENTED** [review]
This SQL looks focused on learning platform data, so I would encourage @code-dot-org/learning-platform to look at the details and sign off that this code is properly getting the data it expects. (That said, a top-level comment describing in more detail what data it expects to get would be very helpful.)

In terms of the goal of creating a scheduled process around this transformation, my recommendation would be to consider a [materialized view](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-overview.html) for this if possible- that takes place entirely within Redshift and wouldn't require any additional external infrastructure to create/maintain.

File: aws/redshift/tables/historic_students.sql (hunk 1 above)
Comment (bakerfranke) [reply aws/redshift/tables/historic_students.sql:64]: @wjordan thanks for the note about a materialized view.  TIL.  If a materialized view is a significant enough performance improvement we may like to go that route.  @Sam-Hutchins what do you think?

Comment (Sam-Hutchins) [reply aws/redshift/tables/historic_students.sql:64]: @wjordan my understanding was that the scheduling process would take place in a system outside of redshift.

**Comment by Sam-Hutchins** [comment]
> In terms of the goal of creating a scheduled process around this transformation, my recommendation would be to consider a [materialized view](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-overview.html) for this if possible- that takes place entirely within Redshift and wouldn't require any additional external infrastructure to create/maintain.

From what I can tell, we still won't have a way to automatically schedule the MV to update without the addition of a new process.  The goal here is to align these tables with what is in dashboard_production.users when redshift updates daily.

**Comment by Sam-Hutchins** [comment]
Sorry, unintentionally closed the PR.

**Pull request closed without merging**